### PR TITLE
Include SCSI sense support and fix regression of SES functions.

### PR DESCRIPTION
**Pull request by cathay4t**
Fix bug: https://bugzilla.redhat.com/show_bug.cgi?id=1360177
Fix bug: https://bugzilla.redhat.com/show_bug.cgi?id=1361107

Changes:
 * Check SCSI sense for error. In above bug, ioctl still return 0 even when certain VPD is not supported.
 * Include SCSI sense information in error message.
 * Fix regression of SES functions caused by f786fa4a0dcb4afed1cf48f4b7780ed73956af8b

### CODE REVIEW

**Comment by cathay4t** [comment]
`local-disk-ident-led-on` failed. Investigating.

**Comment by cathay4t** [comment]
Reopen. The `local-disk-ident-led-on` failure is caused by this patch set. Please continue the code review.

**Comment by cathay4t** [comment]
@joehandzik Could you test the SES functions on hpsa SES with patch set in this PR? Actually, all SES functions in 1.3.2 does not works, we will do a 1.3.3 release to include this PR. Thank you and sorry for the mess.

**Comment by joehandzik** [comment]
@cathay4t Sure, but...any idea when it all broke? I tested successfully, then as far as I knew all that changed after my testing was a bit of refactoring...I'll test better on my end going forward too, now that I'll be building the Debian packages.

**Comment by cathay4t** [comment]
@joehandzik After f786fa4 commit, detail is explained in patch comments.

**Comment by tasleson** [comment]
@cathay4t I don't believe f786fa4a0dcb4afed1cf48f4b7780ed73956af8b caused the regression, it's more of it ensuring that the function returns a null terminated string in all cases which happened to cause an issue because the string wasn't long enough.

File: c_binding/lsm_local_disk.c [hunk 1]
```diff
@@ -781,19 +782,20 @@ static int _sysfs_sas_addr_get(const char *blk_name, char *tp_sas_addr)
         goto out;
 
     sprintf(sysfs_sas_path, "/sys/block/%s/device/sas_address", blk_name);
+    if (! _file_exists(sysfs_sas_path))
+        goto out;
+
+    tmp_rc = _read_file(sysfs_sas_path, (uint8_t *) sysfs_sas_addr, &read_size,
+                        _SYSFS_SAS_ADDR_LEN);
+    /* As sysfs entry has trailing '\n', we should get EFBIG here */
```
Comment (tasleson) [inline c_binding/lsm_local_disk.c]: @cathay4t I don't understand this comment, please elaborate.

**Comment by tasleson** [comment]
@cathay4t Quite a few coverity warnings, please take a look.
```
Error: CPPCHECK_WARNING (CWE-843):
libstoragemgmt-1.3.2/c_binding/libsg.c:281: error[sprintfOverlappingData]: Undefined behavior: Variable 'err_msg' is used as parameter and destination in s[n]printf().
#  279|                                        page_code, ioctl_errno,
#  280|                                        strerror_r(ioctl_errno, strerr_buff,
#  281|->                                                 _LSM_ERR_MSG_LEN), err_msg);
#  282|                       goto out;
#  283|                   } else {

Error: CPPCHECK_WARNING (CWE-843):
libstoragemgmt-1.3.2/c_binding/libsg.c:293: error[sprintfOverlappingData]: Undefined behavior: Variable 'err_msg' is used as parameter and destination in s[n]printf().
#  291|                   _lsm_err_msg_set(err_msg,
#  292|                                    "BUG: Unexpected failure of _sg_io_vpd(): %s",
#  293|->                                  err_msg);
#  294|               }
#  295|           }

Error: FORWARD_NULL (CWE-476):
libstoragemgmt-1.3.2/c_binding/libsg.c:479: assign_zero: Assigning: "sense_fixed" = "NULL".
libstoragemgmt-1.3.2/c_binding/libsg.c:509: var_deref_op: Dereferencing null pointer "sense_fixed".
#  507|           sense_dp = (struct _sg_t10_sense_dp *) sense_data;
#  508|           *sense_key = sense_dp->sense_key;
#  509|->         len = sense_fixed->len + _T10_SPC_SENSE_DATA_LEN_OFFSET;
#  510|           asc = sense_dp->asc;
#  511|           ascq = sense_dp->ascq;

Error: CLANG_WARNING:
libstoragemgmt-1.3.2/c_binding/libsg.c:509:15: warning: Access to field 'len' results in a dereference of a null pointer (loaded from variable 'sense_fixed')
#        len = sense_fixed->len + _T10_SPC_SENSE_DATA_LEN_OFFSET;
#              ^
libstoragemgmt-1.3.2/c_binding/libsg.c:690:11: note: Calling '_sg_io_vpd'
#    _good(_sg_io_vpd(err_msg, fd, _SG_T10_SPC_VPD_DI, vpd_di_data), rc, out);
#    ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
libstoragemgmt-1.3.2/c_binding/utils.h:30:26: note: expanded from macro '_good'
#                rc_val = rc; \
#                         ^
libstoragemgmt-1.3.2/c_binding/libsg.c:259:5: note: Taking true branch
#    if (ioctl_errno != 0) {
#    ^
libstoragemgmt-1.3.2/c_binding/libsg.c:260:9: note: Taking false branch
#        if (page_code == _SG_T10_SPC_VPD_SUP_VPD_PGS) {
#        ^
libstoragemgmt-1.3.2/c_binding/libsg.c:264:13: note: Calling '_check_sense_data'
#        if (_check_sense_data(err_msg, sense_data, &sense_key) != 0) {
#            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
libstoragemgmt-1.3.2/c_binding/libsg.c:479:5: note: 'sense_fixed' initialized to a null pointer value
#    struct _sg_t10_sense_fixed *sense_fixed = NULL;
#    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
libstoragemgmt-1.3.2/c_binding/libsg.c:496:5: note: Control jumps to 'case 115:'  at line 506
#    switch(sense_hdr->response_code) {
#    ^
libstoragemgmt-1.3.2/c_binding/libsg.c:509:15: note: Access to field 'len' results in a dereference of a null pointer (loaded from variable 'sense_fixed')
#        len = sense_fixed->len + _T10_SPC_SENSE_DATA_LEN_OFFSET;
#              ^~~~~~~~~~~
#  507|           sense_dp = (struct _sg_t10_sense_dp *) sense_data;
#  508|           *sense_key = sense_dp->sense_key;
#  509|->         len = sense_fixed->len + _T10_SPC_SENSE_DATA_LEN_OFFSET;
#  510|           asc = sense_dp->asc;
#  511|           ascq = sense_dp->ascq;

Error: COMPILER_WARNING:
libstoragemgmt-1.3.2/c_binding/libsg.c: scope_hint: In function '_sg_io_recv_diag'
libstoragemgmt-1.3.2/c_binding/libsg.c:613:9: warning: too many arguments for format [-Wformat-extra-args]
#         _lsm_err_msg_set(err_msg, "Got error from SGIO RECEIVE_DIAGNOSTIC "
#         ^
#  611|           _check_sense_data(err_msg, sense_data, &sense_key);
#  612|   
#  613|->         _lsm_err_msg_set(err_msg, "Got error from SGIO RECEIVE_DIAGNOSTIC "
#  614|                            "for page code 0x%02x: error %d(%s), msg", page_code,
#  615|                            ioctl_errno,

Error: CPPCHECK_WARNING (CWE-843):
libstoragemgmt-1.3.2/c_binding/libsg.c:617: error[sprintfOverlappingData]: Undefined behavior: Variable 'err_msg' is used as parameter and destination in s[n]printf().
#  615|                            ioctl_errno,
#  616|                            strerror_r(ioctl_errno, strerr_buff, _LSM_ERR_MSG_LEN),
#  617|->                          err_msg);
#  618|           goto out;
#  619|       }

Error: CPPCHECK_WARNING (CWE-843):
libstoragemgmt-1.3.2/c_binding/libsg.c:667: error[sprintfOverlappingData]: Undefined behavior: Variable 'err_msg' is used as parameter and destination in s[n]printf().
#  665|                            "for error %d(%s), %s", ioctl_errno,
#  666|                            strerror_r(ioctl_errno, strerr_buff, _LSM_ERR_MSG_LEN),
#  667|->                          err_msg);
#  668|       }
#  669|   

```

File: c_binding/lsm_local_disk.c (hunk 1 above)
Comment (cathay4t) [inline c_binding/lsm_local_disk.c]: Any sysfs entry with contents has extra trailing '\n', but the way we call _read_file() does not want the trailing '\n'. Let me explained with example:

```
[fge@megadeth ~]$ cat /sys/block/sdr/device/sas_address 
0x500015535990a200
```
We have this SAS address sysfs entry with 16 hex digits and `0x` and trailing '\n' and '\0', so the total length is 20. Previously we call `_read_file()` to request only 19 bytes and replace the last '\n' with '\0', but with f786fa4, the `_read_file()` will fail in this usage with `EFBIG`, then similar usage all failed.

This also apply to `_read_file()` in `_ses_sg_paths_get()` which means no SES sg will be chose, hence all SES functions down.

**Comment by cathay4t** [comment]
@tasleson Fixed the coverity warnings. Thank you.

**Comment by tasleson** [comment]
@cathay4t Looks good, thanks Gris

**Pull request merged**